src/DotNetty.Buffers/ByteBufferUtil.cs

```diff
@@ -295,6 +297,149 @@ static int LastIndexOf(IByteBuffer buffer, int fromIndex, int toIndex, byte valu
             return buffer.ForEachByteDesc(toIndex, fromIndex - toIndex, new IndexOfProcessor(value));
         }
 
+        public static IByteBuffer WriteUtf8(IByteBufferAllocator alloc, string value)
+        {
+            // UTF-8 uses max. 3 bytes per char, so calculate the worst case.
+            IByteBuffer buf = alloc.Buffer(Utf8MaxBytes(value));
+            WriteUtf8(buf, value);
+            return buf;
+        }
+
+        public static int WriteUtf8(IByteBuffer buf, string value)
+        {
+            for (;;)
+            {
+                if (buf is AbstractByteBuffer byteBuf)
+                {
+                    byteBuf.EnsureWritable0(Utf8MaxBytes(value));
+                    int written = WriteUtf8(byteBuf, byteBuf.WriterIndex, value, value.Length);
+                    byteBuf.SetWriterIndex(byteBuf.WriterIndex + written);
+                    return written;
+                }
+                else if (buf is WrappedByteBuffer)
+                {
+                    // Unwrap as the wrapped buffer may be an AbstractByteBuf and so we can use fast-path.
+                    buf = buf.Unwrap();
+                }
+                else
+                {
+                    byte[] bytes = Encoding.UTF8.GetBytes(value);
+                    buf.WriteBytes(bytes);
+                    return bytes.Length;
+                }
+            }
+        }
+
+        internal static int WriteUtf8(AbstractByteBuffer buffer, int writerIndex, string value, int len)
+        {
+            int oldWriterIndex = writerIndex;
+
+            // We can use the _set methods as these not need to do any index checks and reference checks.
+            // This is possible as we called ensureWritable(...) before.
+            for (int i = 0; i < len; i++)
+            {
+                char c = value[i];
+                if (c < 0x80)
+                {
+                    buffer._SetByte(writerIndex++, (byte)c);
+                }
+                else if (c < 0x800)
+                {
+                    buffer._SetByte(writerIndex++, (byte)(0xc0 | (c >> 6)));
+                    buffer._SetByte(writerIndex++, (byte)(0x80 | (c & 0x3f)));
+                }
+                else if (char.IsSurrogate(c))
+                {
+                    if (!char.IsHighSurrogate(c))
+                    {
+                        buffer._SetByte(writerIndex++, WriteUtfUnknown);
+                        continue;
+                    }
+                    char c2;
+                    try
+                    {
+                        // Surrogate Pair consumes 2 characters. Optimistically try to get the next character to avoid
+                        // duplicate bounds checking with charAt. If an IndexOutOfBoundsException is thrown we will
+                        // re-throw a more informative exception describing the problem.
```

Comment:
where does this re-throwing happens?

Reply:
Well, if that happens L316 will through buffer index out of range